### ISSUE TITLE
Customization do not work for a subclass

### ISSUE THREAD

**Issue by RomanKernSW** [AutoFixture/AutoFixture#699]
Hello,
i'm currently try to create an complex object that contains some base classes. I would like to define for a specific subclass explicit values, like:

``` c#
Fixture.Customize<BaseResponse>(a => a.
               With(b => b.Exceptions, new string[0]).
               With(b => b.Code, HttpStatusCode.OK).
               With(b => b.Description, HttpStatusCode.OK.ToString()).
               With(b => b.Notifications, new Notification[0])
               );
```

But this values are never created if I do create new Mock data like:

``` c#
var request = Fixture.Create<InternalResponse>()
```

they do always have default values

on the other hand:

``` c#
Fixture.Customize<System.Runtime.Serialization.IExtensibleDataObject>(a => a.With(b => b.ExtensionData, null));
```

works perfectly. All classes that implements this Interface do not have IExtensibleDataObject Property.

if I try to create BaseResponse, all properties are correctly generated as my customization describes. 

BaseResponse is a simple class (not abstract). 

the other customization do not work either.

``` c#
Fixture.Customize<IEmailAddress>(a => a.With(b => b.EmailAdress, $"test{counter++}@acme.com"));
```

The EmailAddress property is not set. The implementation of this interface is done explicitly, because e-mail properties have different names in different classes.
sure, explicit implementation is difficult, because the property name that have been accessed can have different name. my expectation was, that the customization will called after default values has been created, but it's not the case.

Is there some simple **workaround** for this two problem?

**Comment by akamud** [comment]
I am having the same problem. Is this a bug or an oversight from my part reading the cheat sheet?

**Comment by zvirja** [comment]
I've investigated this issue and let me describe what is going on.

As I suspected, the `Customize<T>()` API isn't designed to cover these kind of tasks. Rather, it expect that you will specify the final type instance that will be activated (`InternalResponse` in your case).

The way how this feature works is that it adds a custom graph of builders to handle requests for the customized type. This graph activates the type using a constructor and post-processes the instance if needed (e.g. assign the properties). This graph is triggered only if current request is exactly of specified type, so it doesn't participate when you configure properties for the base class.

I see a few ways to achieve that.

## Custom behavior

[Behavior](https://github.com/AutoFixture/AutoFixture/wiki/Internal-Architecture#behaviors) is a special kind of interceptor that allows to execute code _before_ and _after_ request is processed. We could check if current request is compatible with our filter and if so - assign properties to it.

Here is the sample of such behavior that allows to achieve the desired behavior:
```csharp
public static class PostActionsCustomization
{
    public static IFixture CustomizePostActions<T>(
        this IFixture fixture, 
        Func<PostActionTransformation<T>, ISpecimenBuilderTransformation> customizations)
    {
        fixture.Behaviors.Add(customizations.Invoke(new PostActionTransformation<T>()));

        return fixture;
    }

    public class PostActionTransformation<T> : ISpecimenBuilderTransformation
    {
        private readonly List<ISpecimenCommand> _commands = new List<ISpecimenCommand>();

        public ISpecimenBuilder Transform(ISpecimenBuilder builder)
        {
            return new Postprocessor(
                builder,
                new CompositeSpecimenCommand(_commands),
                new IsAssignableToTypeSpecification(typeof(T)));
        }

        public PostActionTransformation<T> With<TProperty>(Expression<Func<T, TProperty>> propertyPicker,
            TProperty value)
        {
            _commands.Add(new BindingCommand<T, TProperty>(propertyPicker, value));

            return this;
        }
    }

    /// <summary>
    /// Specification that checks that request type is assignable to the specified type
    /// </summary>
    private class IsAssignableToTypeSpecification : IRequestSpecification
    {
        private readonly Type _expectedType;

        public IsAssignableToTypeSpecification(Type expectedType)
        {
            _expectedType = expectedType ?? throw new ArgumentNullException(nameof(expectedType));
        }

        public bool IsSatisfiedBy(object request)
        {
            if (request is Type typeRequest)
            {
                return _expectedType.IsAssignableFrom(typeRequest);
            }

            return false;
        }
    }
}
```

And here is a sample of its usage:
```csharp
Fixture.CustomizePostActions<BaseResponse>(c => c.
               With(x => x.Exceptions, new string[0]).
               With(x => x.Code, HttpStatusCode.OK).
               With(x => x.Description, HttpStatusCode.OK.ToString()).
               With(x => x.Notifications, new Notification[0])
               );
```
How each time the instance of `BaseResponse` type is activated, the specified properties will be assigned to it.

Notice, this solution is not perfect and properties will be assigned multiple times during the object creation - first time during object activation, second time in our behavior (and probably more times if you add a few customizations). Therefore, some business logic might be broken if you don't expect default values.


## Customization for MemberInfo
Another way is to create helpers to easily customize creation of the particular member infos. This way when AutoFixture will try to resolve a particular `PropertyInfo` or `FieldInfo` to assign it to our object, we will return a custom value.

Example of such customization:
```csharp
public static class PropertiesCustomizerExtension
{
    public static IFixture CustomizeProperties<T>(this IFixture fixture, Func<PropertiesCustomizer<T>, ICustomization> customization)
    {
        fixture.Customize(customization.Invoke(new PropertiesCustomizer<T>()));
        
        return fixture;
    }
    
    
    public class PropertiesCustomizer<T>: ICustomization
    {
        private List<ISpecimenBuilder> _builders = new List<ISpecimenBuilder>();

        public PropertiesCustomizer<T> With<TProperty>(Expression<Func<T, TProperty>> propertyPicker, TProperty value)
        {
            var equalRequestSpecification = new EqualRequestSpecification(
                propertyPicker.GetWritableMember().Member,
                new MemberInfoEqualityComparer());
            
            _builders.Add(
                new FilteringSpecimenBuilder(
                    new FixedBuilder(value),
                    equalRequestSpecification));

            return this;
        }

        public void Customize(IFixture fixture)
        {
            fixture.Customizations.Insert(0, new CompositeSpecimenBuilder(_builders));
        }
    }
    
    
    //Copied from https://github.com/AutoFixture/AutoFixture/blob/master/Src/AutoFixture/Kernel/ExpressionReflector.cs
    private static MemberExpression GetWritableMember(this LambdaExpression propertyPicker)
    {
        var me = propertyPicker.Body as MemberExpression;
        if (me == null)
            me = propertyPicker.TryUnwrapConversion();
        if (me == null)
            throw new ArgumentException("The expression's Body is not a MemberExpression. Most likely this is because it does not represent access to a property or field.", nameof(propertyPicker));

        var pi = me.Member as PropertyInfo;
        if (pi != null && pi.GetSetMethod() == null)
        {
            throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, "The property \"{0}\" is read-only.", pi.Name), nameof(propertyPicker));
        }
        return me;
    }
    
    //Copied from https://github.com/AutoFixture/AutoFixture/blob/master/Src/AutoFixture/Kernel/ExpressionReflector.cs
    private static MemberExpression TryUnwrapConversion(
        this LambdaExpression exp)
    {
        var ue = exp.Body as UnaryExpression;
        if (ue != null)
            return ue.Operand as MemberExpression;
        else
            return null;
    }
}
```
Usage
```csharp
Fixture.CustomizeProperties<BaseResponse>(c => c.
               With(x => x.Exceptions, new string[0]).
               With(x => x.Code, HttpStatusCode.OK).
               With(x => x.Description, HttpStatusCode.OK.ToString()).
               With(x => x.Notifications, new Notification[0])
               );
```

This way looks more natural compared to the Behaviors approach. Also property will be assigned once only.
The drawback is that it doesn't work for explicitly implemented properties, as AutoFixture doesn't assign them by default.

So it's up to you to decide which way suits better for your solution.

## Magic explanation

Another thing I want to mention is why the following code seems to work:
```csharp
Fixture.Customize<System.Runtime.Serialization.IExtensibleDataObject>(a => a.With(b => b.ExtensionData, null));
```

Actually, it doesn't work entirely and if you try to assign anything different from null - you will not receive that value. When you customize a particular property, AutoFixture is configured to not longer resolve that particular property automatically, so it will be explicitly assigned the with configured value later. It works similarly to my "Customization" approach above, with difference that nothing is returned instead of value.

Later, when you create an `InternalRequest` and kernel tries to resolve the `ExtensionData` property value, it skips it. As result that property is not assigned and you see a null.

Therefore, I'd say that the observed behavior it's rather a side effect than a feature.

Let me know if you have more questions to clarify 😉

**Comment by akamud** [comment]
Great in-depth explanation, thanks for taking the time writing this.

**Comment by zvirja** [comment]
The "Magic explanation" part has been already solved via #781 and the fix will be released as a part of v4. That will help to make it less confusing as there will not be an illusion that AF handles such cases somehow.

I'm closing this one as it doesn't seem that more actions are required here. Feel free to re-open the issue if needed.

**Comment by wertzui** [comment]
@zvirja Your code is brilliant, exactly what I was looking for. Thanks for that! :D

Would you mind if I modify it a bit (I needed value factories instead of constant values) and turn it into a NuGet Package?
